pkg/extensions/tbtc/tbtc_test.go

```diff
@@ -0,0 +1,428 @@
+package tbtc
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math/rand"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/keep-network/keep-ecdsa/pkg/chain/local"
+)
+
+const (
+	timeout        = 500 * time.Millisecond
+	depositAddress = "0xa5FA806723A7c7c8523F33c39686f20b52612877"
+)
+
+func TestRetrievePubkey_TimeoutElapsed(t *testing.T) {
+	ctx := context.Background()
+	tbtcChain := local.NewTBTCLocalChain()
+	tbtc := &tbtc{tbtcChain}
+
+	err := tbtc.monitorRetrievePubKey(
+		ctx,
+		constantBackoff,
+		timeout,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tbtcChain.CreateDeposit(depositAddress)
+
+	keepPubkey, err := submitKeepPublicKey(depositAddress, tbtcChain)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait a bit longer than the monitoring timeout
+	// to make sure the potential transaction completes
+	time.Sleep(2 * timeout)
+
+	expectedRetrieveSignerPubkeyCalls := 1
+	actualRetrieveSignerPubkeyCalls := tbtcChain.Logger().RetrieveSignerPubkeyCalls()
+	if expectedRetrieveSignerPubkeyCalls != actualRetrieveSignerPubkeyCalls {
+		t.Errorf(
+			"unexpected number of RetrieveSignerPubkey calls\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]",
+			expectedRetrieveSignerPubkeyCalls,
+			actualRetrieveSignerPubkeyCalls,
+		)
+	}
+
+	depositPubkey, err := tbtcChain.DepositPubkey(depositAddress)
+	if err != nil {
+		t.Errorf("unexpected error while fetching deposit pubkey: [%v]", err)
+	}
+
+	if !bytes.Equal(keepPubkey[:], depositPubkey) {
+		t.Errorf(
+			"unexpected public key\n"+
+				"expected: [%v]\n"+
+				"actual:   [%v]",
+			keepPubkey,
+			depositPubkey,
+		)
+	}
+}
+
+func TestRetrievePubkey_StopEventOccurred(t *testing.T) {
+	ctx := context.Background()
+	tbtcChain := local.NewTBTCLocalChain()
+	tbtc := &tbtc{tbtcChain}
+
+	err := tbtc.monitorRetrievePubKey(
+		ctx,
+		constantBackoff,
+		timeout,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tbtcChain.CreateDeposit(depositAddress)
+
+	keepPubkey, err := submitKeepPublicKey(depositAddress, tbtcChain)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait a while before triggering the stop event because the
+	// extension must have time to handle the start event
+	time.Sleep(100 * time.Millisecond)
+
+	err = tbtcChain.RetrieveSignerPubkey(depositAddress)
```

Comment:
Let's add a comment that this call is a "stop piece" in this test. It took me a while to figure it out why we are calling it and not leaving it up to `monitorRetrievePubKey` 😆